### ISSUE TITLE
Cabal fails on non-breaking space (U+00A0) in .cabal file when using GHC 9.2.1 on M1 Mac

### ISSUE THREAD

**Issue by Gabriella439** [haskell/cabal#7881]
**Describe the bug**

Cabal fails on a `.cabal` file containing a non-breaking space when `Setup.hs` is built using GHC 9.2.1

I found this issue in the wild when building `hiedb-0.4.1.0` using GHC 9.2.1.  The `hiedb.cabal` file contains a non-breaking space right before the `==9.0.1` in the `tested-with` lines

The output I get when running `cabal configure --verbose` on that file is:

```
Using Parsec parser
not implemented 239
CallStack (from HasCallStack):
  error, called at libraries/Cabal/Cabal/src/Distribution/Parsec/FieldLineStream.hs:60:19 in Cabal-3.6.0.0:Distribution.Parsec.FieldLineStream
```

**To Reproduce**

I minimized the reproduction to the following Haskell project with two files:

https://gist.github.com/Gabriel439/950bfb2d5250304c9ad7ae5a7c42d119

```ShellSession
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.2.1

$ ghc Setup.hs

$ ./Setup --version
Cabal library version 3.6.0.0

$ ./Setup configure --verbose
Using Parsec parser
not implemented 239
CallStack (from HasCallStack):
  error, called at libraries/Cabal/Cabal/src/Distribution/Parsec/FieldLineStream.hs:60:19 in Cabal-3.6.0.0:Distribution.Parsec.FieldLineStream
```

**Expected behavior**

The non-breaking space in the `hiedb.cabal` file was probably unintentionally added, but I'm not sure if it should be permitted by Cabal or not since the `.cabal` file format doesn't have an official grammar as far as I can tell.  However, I double-checked and the file is definitely valid UTF-8, so it shouldn't be failing due to a UTF-8 decoding error.

However, the way in which this fails is unexpected and likely indicative of an underlying bug in GHC 9.2.1.  Specifically, the error message indicates that this fails when `c0` is `239` (`0xEF`) within the internal `Distribution.Parsec.FieldLineStream.unconsChar` function, but that case is already handled by a prior guard:

https://github.com/haskell/cabal/blob/9d9fe65d1e6db56004a00f1908207d5ea4ed18da/Cabal/src/Distribution/Parsec/FieldLineStream.hs#L56-L60

… so the `error` should not have been triggered.  Indeed, compiling `Setup.hs` with GHC 8.10.7 does not exhibit this problem and the `configure` step succeeds.

The reason I'm still reporting this issue in the Cabal project is mainly for discoverability purposes, so if somebody runs into this same issue they can find this report.  Once I can narrow down the underlying GHC bug I'll file an upstream bug report and link it to here.

**System information**

 - Operating system: macOS 12.0.1
 - GHC version: 9.2.1
 - Cabal version: 3.6.0.0

**Comment by Mikolaj** [comment]
Thank you for the report. Let's not close until it's fixed in GHC and the buggy GHCs are no longer widely used.

**Comment by Bodigrim** [comment]
https://github.com/haskell/cabal/blob/9d9fe65d1e6db56004a00f1908207d5ea4ed18da/Cabal/src/Distribution/Parsec/FieldLineStream.hs#L57-L59

FWIW these three lines are nonsence, UTF-8 does not allow such sequences.

I would strongly suggest Cabal to use proper `text` facilities instead of inventing its own UTF-8 decoder, which is likely to be both slower and more error-prone.

**Comment by Mikolaj** [comment]
@Bodigrim; great find! BTW, congrats on the UTF8 text release.

I wonder if `base (>=4.9 && <5)` from that release matches our GHC backward compatibility requirements/practice. BTW, I see we already depend on 

    text         (>= 1.2.3.0  && < 1.3) || (>= 2.0 && < 2.1),

in Cabal, so it's the matter of nuking the first alternative and using the newfound power.

**Comment by phadej** [comment]
> I wonder if base (>=4.9 && <5) from that release matches our GHC backward compatibility requirements/practice. BTW, I see we already depend on

It doesn't, (empasis mine)

> Our GHC support window is five years for the Cabal library and three years for cabal-install: that is, the Cabal library must be buildable out-of-the-box **with the dependencies that shipped with GHC** for at least five years.

**Comment by Mikolaj** [comment]
Thanks. The good news is that in 6 years we should be able to use the UTF8 text.

**Comment by Gabriel439** [comment]
Correct me if I'm wrong, but it seems like the fix proposed by @Bodigrim does not depend on `text-2.*`.  All versions of the `text` package support decoding UTF8 (even if the internal representation is UTF-16 for older versions)

**Comment by Mikolaj** [comment]
Not sure, but is the UTF8 handling in text from 5 years (!) ago better than what's in that code snippet?

**Comment by Bodigrim** [comment]
> Not sure, but is the UTF8 handling in text from 5 years (!) ago better than what's in that code snippet?

Yes, it is.

**Comment by phadej** [comment]
That code existed in Cabal codebase for 12 or so years, https://github.com/haskell/cabal/commit/ca7262e969 where it was borrowed from utf8-string, which seems to have similar implementation https://hackage.haskell.org/package/utf8-string-1.0.2/docs/src/Codec.Binary.UTF8.String.html#decode 

The `otherwise = error $ "not implemented " ++ show c0` clause is there because that code SHOULD only be called with valid UTF-8, so we don't replace, but fail hard.

I'd be happy to know why code compiled with GHC-8.10.7  works, but with GHC-9.2.1 doesn't, before we start to swap implementations.

**Comment by phadej** [comment]
I also cannot reproduce this. Using GHC-9.2.1, I `cabal get hiedb-0.4.1.0`,

```
% ghci-9.2.1
GHCi, version 9.2.1: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/phadej/.ghci
Prelude> :m +Distribution.Verbosity Distribution.PackageDescription.Parsec 
Prelude Distribution.Verbosity Distribution.PackageDescription.Parsec> readGenericPackageDescription silent "hiedb.cabal" 
```

@Gabriel439 Are you using M1 mac, by accident?

**Comment by Bodigrim** [comment]
> The `otherwise = error $ "not implemented " ++ show c0` clause is there because that code SHOULD only be called with valid UTF-8, so we don't replace, but fail hard. 

In other cases Cabal implementation replaces invalid UTF-8 sequences with `replacementChar` instead of failing. Sigh.

**Comment by phadej** [comment]
`text` has `decodeUtf8` which also throws. A good sanity check when you know about it. I hope that `text` also throws when it realizes its internal buffer is in invalid state. (E.g. in `Data.Text.uncons`!)

**Comment by Gabriel439** [comment]
Yes, I'm using an M1 Mac (`aarch64-darwin`)

**Comment by Gabriel439** [comment]
> Not sure, but is the UTF8 handling in text from 5 years (!) ago better than what's in that code snippet?

Yes, `text-1.2.3.0` uses a more efficient implementation than that code snippet.  See:

https://hackage.haskell.org/package/text-1.2.3.0/src/cbits/cbits.c

**Comment by andreasabel** [comment]
> I also cannot reproduce this. 
> @Gabriel439 Are you using M1 mac, by accident?

This problem is not reproducible on my x86_64 mac.

**Comment by wz1000** [comment]
Is there a GHC ticket tracking this?

**Comment by Bodigrim** [comment]
That's most likely a known issue with GHC 9.2.1 being broken on ARM because of subword primops.

**Comment by phadej** [comment]
@wz1000 yes, see https://discourse.haskell.org/t/psa-9-2-1-aarch64-miscompilation/3638

IMO this issue can be closed, as it looks like it's a GHC thing. Or we can wait for GHC-9.2.2 to verify it's actually that. I don't think we need to to anything in the meantime though.

**Comment by ulysses4ever** [comment]
Should this be closed? And perhaps should there be a new ticket about replacing UTF8-decoding stuff with a `text`-based implementation?

**Comment by Gabriella439** [comment]
I think it's fine to close it.  I only opened this for discoverability reasons, so that people searching would find something in the issue tracker.  I'll defer to the maintainers' judgment

**Comment by andreasabel** [comment]
@Gabriella439 : Is the OP fixed for you with GHC >= 9.2.2 ?

**Comment by Gabriella439** [comment]
Yes, this is fixed for me when upgrading to GHC 9.2.2, so I'll mark this closed.
